idseq_pipeline/commands/assembly_functions.py

```
@@ -0,0 +1,134 @@
+        return output, sorted_taxids_to_assemble
+
+    def length_without_newlines(sequence):
+        return len(sequence.replace("\n",""))
+
+    def read_fasta(input_fasta):
+        ''' Read fasta file into dictionary. Newlines remain part of the keys/values. '''
+        read_name = None
+        d = {}
+        read_order = []
+        with open(input_fasta, 'rb') as f:
+            for line in f:
+                if line.startswith('>'):
+                    read_name = line
+                    read_order += [read_name]
+                else:
+                    d[read_name] = d.get(read_name, '') + line
+        return d, read_order
+
+    def write_fasta(fasta_dict, read_order, output_file):
+        with open(output_file, 'wb') as f:
+            for read_name in read_order:
+                f.write(read_name + fasta_dict[read_name])
+
+    def clean_scaffolds(spades_scaffolds_fasta, min_contig_length, cleaned_fasta):
+        fasta_dict, read_order = read_fasta(spades_scaffolds_fasta)
+        for name, sequence in fasta_dict.iteritems():
+            if length_without_newlines(sequence) <= min_contig_length:
+                read_order.remove(name)
+        write_fasta(fasta_dict, read_order, cleaned_fasta)
+
+    def max_read_length(input_fasta):
+        fasta_dict, _read_order = read_fasta(input_fasta)
+        return max([length_without_newlines(sequence) for sequence in fasta_dict.values()])
+
+    def spades(input_fasta, output_fasta):
+        tmp_output_dir = input_fasta + "_temp_output"
+        tmp_output_file = tmp_output_dir + "/scaffolds.fasta"
+        try:
+            execute_command_realtime_stdout("spades.py -s %s -o %s -m 60 -t 32 --only-assembler" % (input_fasta, tmp_output_dir))
+            subprocess.check_call("mv %s %s" % (tmp_output_file, output_fasta), shell=True)
+            return True
+        except:
+            return False
```

When you swallow an exception it's usually good to print the stacktrace in the log, i.e., `traceback.print_exc()`.

---

It probably won't be super informative in this case.

---

ok!